src/spec-common/variableSubstitution.ts

```diff
@@ -32,6 +33,11 @@ export function substitute<T extends object>(context: SubstitutionContext, value
 	return substitute0(replace, value);
 }
 
+export function beforeContainerSubstitute<T extends object>(idLabels: Record<string, string>, value: T): T {
+	let devcontainerId: string | undefined;
+	return substitute0(replaceDevContainerId.bind(undefined, () => devcontainerId || (devcontainerId = devcontainerIdForLabels(idLabels))), value);
```

Comment:
Curious, would both code result into same output ? 

```suggestion
	return substitute0(replaceDevContainerId.bind(undefined, () => devcontainerId || devcontainerIdForLabels(idLabels))), value);
```

Reply:
The output would be the same, but the variable `devcontainerId` wouldn't be used for caching because it would not get the id assigned when it is computed the first time. Performance would be worse when `${devcontainerId}` is used multiple times.